### ISSUE TITLE
Java version detection

### ISSUE THREAD

**Issue by Tristan971** [dorkbox/SystemTray#90]
The current `dorkbox.util.OS#_getJavaVersion` does not support versions higher than Java 9 (e.g. Java 10).

This breaks the macOS + JavaFX compatibility (the system assumes we are running a JDK lower than 8 and thus can't use the system tray on macOS).

Cannot make a PR due to dead links regarding the Utilities project this one depends on.
Will bake some implementation suggestion later though.

**Comment by tresf** [comment]
Ran into a similar problem downstream with Semver parsing.  Here was my solution... https://github.com/qzind/tray/commit/49e553046621114b14787297b5a9eb17efc63802#diff-e3b46f596e3185d1f81105720591194eR44

**Comment by Tristan971** [comment]
That's a clever way to go but I would wager that if we can find the .class binary version we can probably more easily build on top of that and forever decouple from almost-marketing-ish issues? idk really haha

**Comment by tresf** [comment]
> marketing-ish issues

Yeah, in hindsight, the problem was the everlasting 1.x.x versioning, so they've fixed that with [JEP 223](http://openjdk.java.net/jeps/223).  I agree, `1.9.0` is almost useless, but it allows version comparisons.

If you want to inspect the class file, you'll be getting into [some strangeness](https://stackoverflow.com/a/1096159/3196753) with cross-lookups to translate back to human-readable versions.

Our downstream project actually cares about the `uXXX` version as well so since that commit, we've [added support for the buildMetaData](https://github.com/qzind/tray/commit/0d91e06f8a7a0348d64b6066cf9d22d7b8b4beb6#diff-e3b46f596e3185d1f81105720591194eR70).  This is important when we get a specific patch from Oracle that arrives mid-release and we can remove workaround code -- something not yet parsed from the JDK9 version info.

**Comment by Tristan971** [comment]
Yeah unfortunately despite Java 10 fixing that, I'm pretty sure every lib out there will be supporting *at least* Java 8 until something like 2030 because... yeah... So if your last comment's proposition works, that would be pretty sweet actually.
Although imo this does not belong in Dorkbox/SystemTray but in its own library which I'm sure already exists somewhere?

**Comment by dorkbox** [comment]
I like the approach checking class version numbers.

Regrettably, I did not do think past Java9 when I wrote that.  

Utilities isn't dead, it's just moved (SystemTray will also be moved, just not yet), and when the build system is done, everything should all work together nicely.

**Comment by Tristan971** [comment]
Yeah in the end class file version probably is a biit too flaky to be sure of the executing JVM and we'd need to get access to running source code and not the lib's source code itself so it might not be right... Unsure how to go about all that. Will see if I find a lib.

**Comment by dorkbox** [comment]
Keep us updated :) If not, I'll just patch up and fork jsemver to do it.

**Comment by Tristan971** [comment]
Tbh if I don't find anything I'm building it hahaha

**Comment by tresf** [comment]
Apache's awesome, but that's about the ugliest thing I've ever seen.  Any library that has to be patched that often is ticking time bomb. :)

**Comment by dorkbox** [comment]
What about a hybrid, so instead of passing in `JavaVersion.JAVA_9` which would need a new update for every release, and change the parameter to an `int` - so something like `SystemUtils.isJavaVersionAtLeast(9)`

Then it's solving the problems for Java < 9 and Java >= 9, as long as the specification version property format stays the same?

**Comment by Tristan971** [comment]
That's true, although I think that if something is ever going to be patched it would be apache commons haha. But yeah I can see why you would not like it.
Guess it's off to the races to build a decent parser then.

**Comment by tresf** [comment]
> Then it's solving the problems for Java < 9 and Java >= 9, as long as the specification version property format stays the same?

Which is what every SemVer library does. :)

> Guess it's off to the races to build a decent parser then.

👍

**Comment by dorkbox** [comment]
> Which is what every SemVer library does. :)

Huh. TIL. I'll have to take a closer look because `1.1` and `10` violate the heck out of semver.org (which is why I already forked jsemver for myself to permit the exclusion of patch version information... I don't use patch version numbers unless there is a patch version release, but also don't want to rewrite version validation, comparison, incrementing, decrementing, parsing, etc. jsemver did a good job.)

> Guess it's off to the races to build a decent parser then.

I'll just update `dorkbox.util.OS#_getJavaVersion` to use what jsemver did :)

**Comment by Tristan971** [comment]
Haha alright then, still sounds like something could/should have been made about that in general :)

**Comment by dorkbox** [comment]
yea, i totally agree. I'll update Utilities tonight to have the correct version parsing stuff from jsemver, so the next build of SystemTray will be correct with java version checking.

**Comment by Tristan971** [comment]
Perfect, curious about utilities though, what is the correct git repo url?

**Comment by dorkbox** [comment]
https://git.dorkbox.com/dorkbox/Utilities

**Comment by dorkbox** [comment]
Or rather, i should say: git@git.dorkbox.com:dorkbox/Utilities.git
